src/Compilers/Core/Portable/Compilation/DeterministicKey.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class SyntaxTreeKey
+    {
+        public abstract string FilePath { get; }
+        public abstract ParseOptions Options { get; }
+
+        protected SyntaxTreeKey()
+        {
+        }
+
+        public abstract SourceText GetText(CancellationToken cancellationToken = default);
+
+        public static SyntaxTreeKey Create(SyntaxTree tree)
+            => new DefaultSyntaxTreeKey(tree);
+
+        private sealed class DefaultSyntaxTreeKey : SyntaxTreeKey
+        {
+            private readonly SyntaxTree _tree;
+
+            public DefaultSyntaxTreeKey(SyntaxTree tree)
+            {
+                _tree = tree;
+            }
+
+            public override string FilePath
+                => _tree.FilePath;
+
+            public override ParseOptions Options
+                => _tree.Options;
+
+            public override SourceText GetText(CancellationToken cancellationToken = default)
+                => _tree.GetText(cancellationToken);
+        }
+    }
+
+    internal static class DeterministicKey
```

Comment:
nit: is there a benefit of separating `DeterministicKey` and `DeterministicKeyBuilder`? I'd be tempted to move those static APIs over into `DeterministicKeyBuilder`.

Reply:
@jaredpar Did you have a though on this?

Reply:
Presently I like the separation because the `DeterministicKey` API is completely stateless. It's a `static class` which has all inputs which are effectively immutable data and returns a key. It fits with the theme of determinism. The builder APIs though are stateful, involve hierachies and non-static types. 

I thought about merging them but then we end up having to expose `abstract class` instances to customers for types they can never instantiate and for which there are no instance methods available. Felt wrong.

Reply:
Alright. Thanks